### DISCUSSION TITLE
Definition and use of recursive sub in Mojo template

### DISCUSSION THREAD

**Discussion by belfasthenry** [mojolicious/mojo#1656] [Q&A] [answered]
I have a template which renders a directory tree.  A separate module scans the disk and builds a hash-based tree structure which represents the structure of directories and sub-directories.  (It also counts the images in each directory but that's not relevant here).  The structure which that module returns is of this form
```
{
  2020 => {
    imagecount => 2,
    April => {},
    May => {
      Mayday => {
        imagecount => 3
      }
    },
  ... etc
};
```
Of course I can't tell how deep the tree goes, so to generate the HTML to display it I need a recursive function.  Following [https://stackoverflow.com/questions/26254966/creating-recursive-template-blocks-in-mojolicious](url) I coded this (omitting a few lines for brevity)
```
use v5.16;                                  # I'm running perl v5.30.0
% my $tree_sub = begin
%   my $tree_data = shift;                  # tree_data is in the stash
%   state $lev = 0;
%   $lev++;
    <ul class="level$lev">
%   for my $k ( sort keys %$tree_data ){
%     next if $k eq 'imagecount';
      <li>$k</li>
%     __SUB__->( $tree_data->{$k} );
%   }
    </ul>
%   $lev--;
% end

<html>
<div class="w3-row">
  <div class="w3-col s3 m2 l2">
    %= $tree_sub->( $tree_hr )
  </div>

... etc
```
But running it under morbo I get `Undefined subroutine &Mojo::Template::Sandbox::0b9aa13c56da7fa1d0c4081edd141164::__SUB__ called at template dirlist/dirlist.html.ep`

That Stack Exchange article is over 6 years old: has something changed since then to invalidate my code?  What am I doing wrongly?

**Comment by jberger** [comment, marked as answer]
I think you're rendering `use v5.16;` to the client rather than using it as a directive. Put a % in front of that line. If that doesn't fix it, put `% use feature 'current_sub';` on the second line

**Reply by belfasthenry** [reply]
I'm tremendously grateful for the speed and accuracy of your responses; thank you.

The 'use feature' produced the required result.  In the interests of completeness I took out the 'use v5.16' statement and it still worked, so for the record the successful (abbreviated) code - running Perl v5.30 - is this:
```
% use feature 'current_sub';
% my $tree_sub = begin
%   my $tree_data = shift;
%   state $lev = 0;
%   return unless ref($tree_data) eq 'HASH';
%   $lev++;
    <ul class="level$lev">
%   for my $k ( sort keys %$tree_data ){
%     next if $k eq 'imagecount';
      <li><%= $k %></li>
%     __SUB__->( $tree_data->{$k} );
... etc
```

**Reply by belfasthenry** [reply]
For the benefit of those who may happen upon this thread when solving a similar problem, the code above still needs modifying.  Calling `__SUB__->( $tree_data->{$k} )` doesn't render the HTML that was produced by the subroutine: it just vanishes into the howling aether.  Instead you should code something like this:
```
%    my $ret =  __SUB__->( $tree_data->{$k} );
      <%=$ret%>
```
